### PR TITLE
Detailed RenderBox and RenderObject docs

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
@abarth

File: packages/flutter/lib/src/rendering/box.dart [hunk 1]
```diff
@@ -527,22 +527,411 @@ class _IntrinsicDimensionsCacheEntry {
...
+/// [RenderProxyBox] assumes that the child is the size of the parent and
+/// positioned at 0,0. If this is not true, then use [RenderShiftedBox] instead.
+///
+/// See
+/// [proxy_box.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/rendering/proxy_box.dart)
+/// for examples of inheriting from [RenderProxyBox].
+///
+/// #### Using RenderShiftedBox
+///
+/// By default, a [RenderShiftedBox] acts much like a [R]enderProxyBox but
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: `[R]` ?

File: packages/flutter/lib/src/rendering/box.dart [hunk 2]
```diff
@@ -527,22 +527,411 @@ class _IntrinsicDimensionsCacheEntry {
...
+/// [sizedByParent] to return true, and then override [performResize] to set the
+/// [size] using nothing but the constraints, e.g.:
+///
+/// ```dart
+/// @override
+/// bool get sizedByParent => true;
+///
+/// @override
+/// void performResize() {
+///   size = constraints.biggest;
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: Do we want to recommend biggest?  That will blow up with infinite constraints.

File: packages/flutter/lib/src/rendering/box.dart [hunk 3]
```diff
@@ -527,22 +527,411 @@ class _IntrinsicDimensionsCacheEntry {
...
+/// whatever the output of [layout] is for that layout protocol. The
+/// `parentUsesSize` flag is still used to indicate whether the parent is going
+/// to read that output, and optimisations still kick in if the child has tight
+/// constraints (as defined by [Constraints.isTight]).
+///
+/// ### Painting
+///
+/// To describe how a render box paints, implement the [paint] method. It is
+/// given a [PaintingContext] object and an [Offset]. The offset specifies the
+/// position that the box should consider as its top left corner. The painting
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: its top left corner...

in the coordinate system of `context.canvas`

File: packages/flutter/lib/src/rendering/object.dart [hunk 4]
```diff
@@ -1058,24 +1100,130 @@ void _doNothing() { }
...
+/// appropriate for your class and passes that to the child's [layout] method.
+///
+/// ### Lifecycle
+///
+/// The lifecycle of a [RenderObject] is roughly as follows:
+///
+/// 1. creation (constructor runs)
+/// 2. properties get modified
+/// 3. layout
+/// 4. paint
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/object.dart]: Semantics?

**Comment by abarth** [comment]
LGTM

**Comment by Hixie** [comment]
Made fixes, will land when Travis agrees.

**Pull request merged**